File: server/app/api/finance.py
"""金融分析 API：概览、单企业分析、Markdown 报告。

数据来自 finance 表（东财财务摘要 + 新浪三表字段级合并），
所有计算在 app/services/finance.py，本层只做参数校验与响应组装。
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import PlainTextResponse
from sqlalchemy.orm import Session as DbSession

from app.db.database import get_db
from app.services import finance as fin

router = APIRouter(prefix="/api/finance", tags=["finance"])


@router.get("/overview")
def overview(years: int = Query(5, ge=1, le=10), db: DbSession = Depends(get_db)):
    """全库财务概览（供选择器与横向筛选）。"""
    return fin.overview(db, years=years)


@router.get("/{enterprise_id}/analysis")
def analysis(
    enterprise_id: int,
    years: int = Query(5, ge=1, le=10),
    peers: bool = Query(True),
    db: DbSession = Depends(get_db),
):
    """单企业金融分析：KPI、趋势、杜邦、Z/F/M 模型、同业对标、异常勾稽。"""
    data = fin.analysis(db, enterprise_id, years=years, with_peers=peers)
    if data.get("error"):
        raise HTTPException(status_code=404, detail=data["error"])
    return data


@router.get("/{enterprise_id}/report", response_class=PlainTextResponse)
def report(
    enterprise_id: int,
    years: int = Query(5, ge=1, le=10),
    db: DbSession = Depends(get_db),
):
    """确定性 Markdown 财务分析报告（叙述性解读由 Agent 工具完成）。"""
    return PlainTextResponse(
        fin.report_markdown(db, enterprise_id, years=years),
        media_type="text/markdown; charset=utf-8",
    )

File: backend/src/api/finance.rs
//! 金融分析接口：概览 / 单企业分析 / Markdown 报告。

use axum::extract::{Path, Query, State};
use axum::http::header;
use axum::response::IntoResponse;
use axum::Json;
use serde::Deserialize;
use serde_json::Value;

use crate::error::{AppError, AppResult};
use crate::services::finance as fin;
use crate::state::AppState;

#[derive(Deserialize)]
pub struct YearsQuery {
    #[serde(default = "default_years")]
    pub years: usize,
    #[serde(default = "default_true")]
    pub peers: bool,
}

fn default_years() -> usize {
    5
}

fn default_true() -> bool {
    true
}

/// GET /api/finance/overview
pub async fn overview(
    State(st): State<AppState>,
    Query(q): Query<YearsQuery>,
) -> AppResult<Json<Value>> {
    Ok(Json(fin::overview(&st.db, q.years.clamp(1, 10))?))
}

/// GET /api/finance/{id}/analysis
pub async fn analysis(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    Query(q): Query<YearsQuery>,
) -> AppResult<Json<Value>> {
    let data = fin::analysis(&st.db, id, q.years.clamp(1, 10), q.peers)?;
    if let Some(err) = data.get("error").and_then(|v| v.as_str()) {
        return Err(AppError::not_found(err));
    }
    Ok(Json(data))
}

/// GET /api/finance/{id}/report —— 确定性 Markdown 报告
pub async fn report(
    State(st): State<AppState>,
    Path(id): Path<i64>,
    Query(q): Query<YearsQuery>,
) -> AppResult<impl IntoResponse> {
    let text = fin::report_markdown(&st.db, id, q.years.clamp(1, 10))?;
    Ok(([(header::CONTENT_TYPE, "text/markdown; charset=utf-8")], text))
}
